source/neuropod/internal/neuropod_tensor.hh

```diff
@@ -155,7 +155,7 @@ public:
 
     NeuropodTensor(const NeuropodTensor &) = delete;
 
-    virtual ~NeuropodTensor() {}
+    virtual ~NeuropodTensor() override {}
```

Comment:
If this is override, it means its Base class already has virtual dtor. In this case you may just remove it completely (as I remember this considered as a best practice that Derived class doesn't specify empty virtual dtor if its Base has virtual dtor already)

Reply:
unrelated, just reminded me some idea I had once and I want to share with you:

you know that STL containers doesn't have virtual dtor (because "virtual" is not free and STL supports algos then classic OOP approach). But because of that "public inheritance" from STL containers introduces problem of resource release in case of "delete pBase".

I understand that it will never happen with current STL but still theoretically it could solve this issue by allowing user to parametrize explicitly "virtual dtor"/"regular dtor":

STL would need to add one more template param:
 
template<..., class BasewithDtor=BaseWithRegularDtor>
class vector: public BaseWithDtor {...}

struct BaseWithRegularDtor{ virtual ~BaseWithRegularDtor() {} }
struct BaseWithVirtualDtor{ virtual ~BaseWithVirtualDtor() {} }

class MyContainer: public std::vector<..., BasewithDtor=BaseWithVirtualDtor> {}

I never used this technic in real project, this design patter makes sense only for libs like STL  

What do you think, would this "design pattern" can be useful theoretically ?

Reply:
Maybe there are situations where it would be useful, but, with the design above, you lose a lot of the benefits of inheritance since `std::vector<..., VirtualDtor>` and `std::vector<...>` are not interchangeable (in terms of dynamic polymorphism). The remaining major benefit I see would be code reuse, but composition would likely be a better solution than inheritance here.

Also, in your implementation above, both `RegularDtor` and `VirtualDtor` define virtual destructors. A "real" implementation of this likely won't be straightforward.

Either way, I don't think it would be useful here because `NeuropodTensor` is an abstract class and therefore doesn't have any "non-virtual" version